pkg/lib/cockpit-components-context-menu.jsx

```diff
@@ -33,83 +33,83 @@
  *  - setText, method which is called when paste is clicked
  *  - parentId, area in which it listens to left button clicks
  */
-export class ContextMenu extends React.Component {
-    constructor() {
-        super();
-        this.state = { visible: false };
-        this._handleContextMenu = this._handleContextMenu.bind(this);
-        this._handleClick = this._handleClick.bind(this);
-    }
-
-    componentDidMount() {
-        const parent = document.getElementById(this.props.parentId);
-        parent.addEventListener('contextmenu', this._handleContextMenu);
-        document.addEventListener('click', this._handleClick);
-    }
-
-    componentWillUnmount() {
-        const parent = document.getElementById(this.props.parentId);
-        parent.removeEventListener('contextmenu', this._handleContextMenu);
-        document.removeEventListener('click', this._handleClick);
-    }
-
-    _handleContextMenu(event) {
-        event.preventDefault();
-
-        this.setState({ visible: true });
+export const ContextMenu = ({ parentId, getText, setText }) => {
+    const [visible, setVisible] = React.useState(false);
+    const [event, setEvent] = React.useState(null);
+    const root = React.useRef(null);
+
+    React.useEffect(() => {
+        const _handleContextMenu = (event) => {
+            event.preventDefault();
+
+            setVisible(true);
+            setEvent(event);
+        };
+
+        const _handleClick = (event) => {
+            if (event && event.button === 0) {
+                const wasOutside = !(event.target.contains === root.current);
+
+                if (wasOutside)
+                    setVisible(false);
+            }
+        };
+
+        const parent = document.getElementById(parentId);
+        parent.addEventListener('contextmenu', _handleContextMenu);
+        document.addEventListener('click', _handleClick);
+
+        return () => {
+            parent.removeEventListener('contextmenu', _handleContextMenu);
+            document.removeEventListener('click', _handleClick);
+        };
+    }, [parentId]);
+
+    React.useEffect(() => {
+        if (!event)
+            return;
 
         const clickX = event.clientX;
         const clickY = event.clientY;
         const screenW = window.innerWidth;
         const screenH = window.innerHeight;
-        const rootW = this.root.offsetWidth;
-        const rootH = this.root.offsetHeight;
+        const rootW = root.current.offsetWidth;
+        const rootH = root.current.offsetHeight;
 
         const right = (screenW - clickX) > rootW;
         const left = !right;
         const top = (screenH - clickY) > rootH;
         const bottom = !top;
 
         if (right) {
-            this.root.style.left = `${clickX + 5}px`;
+            root.current.style.left = `${clickX + 5}px`;
         }
 
         if (left) {
-            this.root.style.left = `${clickX - rootW - 5}px`;
+            root.current.style.left = `${clickX - rootW - 5}px`;
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18017-20221213-083151-ab71eae1-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)